### DISCUSSION TITLE
Doks by default is extremely slow for large websites

### DISCUSSION THREAD

**Discussion by RC1242RC** [thuliteio/doks#859] [Q&A] [answered]
Hey, thanks for this great theme that I've been using for a documentation project.

Unfortunately, if there are a high number of pages (in my case 6000+) I've found that the default flexsearch options result in website which is extremely slow to both build and run.

For example, my 6000 page website would take around 40ms/page to build, and refused to build locally (the process would just crash on my laptop with an i5 and 8GB of ram). In addition, once the site was built each page would take around 5s to load which made the browsing experience pretty tedious.

I found this pretty confusing because tools like web.dev or lighthouse were still giving very high performance scores. Eventually I ended up trying GTmetrix which found the problem for me:

![Screenshot from 2022-08-02 21-56-10](https://user-images.githubusercontent.com/43422297/182472407-e6622506-154e-462a-9c13-c96a970777ee.png)

![Screenshot from 2022-08-02 22-03-33](https://user-images.githubusercontent.com/43422297/182472912-b234ed3c-6595-4718-93ec-d40990602f0a.png)

Turns out most of that 6s block time was being used on a massive 5MB .js script. From what I can tell this seems to be the flexsearch index, where data about all the pages is stored and then searched through. 

I fixed this by going into `assets/js/index.js`, and edited the bit that stores page data:

```
(function(){

  var index = new FlexSearch.Document({`
    tokenize: "forward",
    cache: 100,
    document: {
      id: 'id',
      store: [
        "href", "title", "description"
      ],
      index: ["title", "description", "content"]
    }
  });
```

I just removed `"content"` from the list, since that's the largest part being stored. I also removed the `"content"` section from this part of the code, since it is no longer doing anything:

```
  {{ range $index, $element := $list -}}
    index.add(
      {
        id: {{ $index }},
        href: "{{ .RelPermalink }}",
        title: {{ .Title | jsonify }},
        {{ with .Description -}}
          description: {{ . | jsonify }},
        {{ else -}}
          description: {{ .Summary | plainify | jsonify }},
        {{ end -}}
        content: {{ .Plain | jsonify }}
      }
    );
  {{ end -}}
```

Now my site builds at around 1ms/page and GTmetrix shows it loading in under a second:

![Screenshot from 2022-08-02 22-09-20](https://user-images.githubusercontent.com/43422297/182475812-c1b46e6d-c052-4c08-899e-15895379e9bd.png)

Unfortunately the trade-off here is that the flexsearch index no longer contains the content of the pages, so when you search you are just really searching the page titles.

I feel like this issue is not trivial since Doks is made for hosting documentation, and documentation is often very large. The solution I've got is ok and I can live with it, but Does anyone know if I can bring back the content searching without running into the same crippling speed issue?

Thanks!

p.s. you can find the site at https://rc1242rc.github.io

**Comment by h-enk** [comment, marked as answer]
Did you try tinkering with the [Memory Allocation](https://github.com/nextapps-de/flexsearch#memory-allocation) settings?

See also https://github.com/nextapps-de/flexsearch/issues/215#issuecomment-869559039

**Reply by RC1242RC** [reply]
Hi @h-enk, thanks for the suggestion. I hadn't touched the memory allocation settings but I have now tried the settings recommended for low memory usage:

```
(function(){

  var index = new FlexSearch.Document({
    tokenize: "strict",
    encoder: "extra",
    optimize: true,
    resolution: 1,
    minlength: 3,
    fastupdate: false,
    context: false,
    document: {
      id: 'id',
      store: [
        "href", "title", "description", "permalink"
      ],
      index: ["title", "description", "permalink", "content"]
    }
  });
```

This has reduced the problem, but not dramatically. I've removed quite a lot of pages from my website and now the time to load each page is around 2s, which is not perfect but workable!
